guides/10_other-tutorials/building-an-mcp-client-with-gradio.md

```diff
@@ -0,0 +1,489 @@
+                            
+                            # Return image info
+                            return json.dumps({
+                                "type": "image",
+                                "url": image_url,
+                                "message": f"Generated image for prompt: {prompt}"
+                            })
+                except:
+                    continue
+        
+        # Wait before polling again
+        time.sleep(2)
+    
+    return json.dumps({
+        "type": "error",
+        "message": "Timeout reached waiting for image generation."
+    })
+
+if __name__ == "__main__":
+    # Initialize and run the server
+    mcp.run(transport='stdio')
+```
+
+### What this server does:
+
+1. It creates an MCP server that exposes a `generate_image` tool
+2. The tool connects to the SanaSprint model hosted on HuggingFace Spaces
+3. It handles the asynchronous nature of image generation by polling for results
+4. When an image is ready, it returns the URL in a structured JSON format
+
+## Part 2: Building the MCP Client with Gradio
+
+Now let's create a Gradio chat interface as MCP Client that connects Claude to our MCP server.
+
+Create a file named `app.py`:
+
+```python
+import asyncio
+import os
+import json
+from typing import List, Dict, Any, Union
+from contextlib import AsyncExitStack
+
+import gradio as gr
+from gradio.components.chatbot import ChatMessage
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+from anthropic import Anthropic
+from dotenv import load_dotenv
+
+load_dotenv()  # Load environment variables from .env
+
+# Create event loop for async operations
+loop = asyncio.new_event_loop()
+asyncio.set_event_loop(loop)
+
+class MCPClientWrapper:
+    """Minimal MCP client wrapper for Gradio"""
+    
+    def __init__(self):
+        self.session = None
+        self.exit_stack = None
+        self.anthropic = Anthropic()
+        self.tools = []
+    
+    def connect(self, server_path: str) -> str:
+        """Connect to the MCP server"""
+        return loop.run_until_complete(self._connect(server_path))
+    
+    async def _connect(self, server_path: str) -> str:
+        """Internal async method to connect to server"""
+        if self.exit_stack:
+            await self.exit_stack.aclose()
+        
+        self.exit_stack = AsyncExitStack()
+        
+        is_python = server_path.endswith('.py')
+        command = "python" if is_python else "node"
+        
+        server_params = StdioServerParameters(
+            command=command,
+            args=[server_path],
+            env={"PYTHONIOENCODING": "utf-8", "PYTHONUNBUFFERED": "1"}
+        )
+        
+        stdio_transport = await self.exit_stack.enter_async_context(stdio_client(server_params))
+        self.stdio, self.write = stdio_transport
+        
+        self.session = await self.exit_stack.enter_async_context(ClientSession(self.stdio, self.write))
```

Comment:
What's the benefit of using `AsyncExitStack` over just using `ClientSession` and `stdio_client` in an `async with` ?

Reply:
Not sure, will leave it to @yvrjsharma to make the change if needed